cranelift/codegen/src/isa/aarch64/lower.isle

```diff
@@ -513,9 +513,21 @@
 
 ;;;; Rules for `fma` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-(rule (lower (has_type (ty_scalar_float ty) (fma x y z)))
-       (fpu_rrrr (FPUOp3.MAdd) (scalar_size ty) x y z))
+(type IsFneg (enum (Result (negate u64) (value Value))))
 
+(decl pure is_fneg (Value) IsFneg)
+(rule 1 (is_fneg (fneg x)) (IsFneg.Result 1 x))
+(rule 0 (is_fneg x) (IsFneg.Result 0 x))
+
+(decl fmadd_series (Type u64 Value Value Value) InstOutput)
+(rule 0 (fmadd_series (ty_scalar_float ty) 0 x y z) (fpu_rrrr (FPUOp3.MAdd) (scalar_size ty) x y z))
+(rule 0 (fmadd_series (ty_scalar_float ty) 1 x y z) (fpu_rrrr (FPUOp3.MSub) (scalar_size ty) x y z))
+
+(rule (lower (has_type (ty_scalar_float ty) (fma x_src y_src z)))
+  (if-let (IsFneg.Result neg_x x) (is_fneg x_src))
+  (if-let (IsFneg.Result neg_y y) (is_fneg y_src))
+  (fmadd_series ty (u64_xor neg_x neg_y) x y z))
```

Comment:
I think that this can be a regular `let`, as `fneg` will never fail, right?

Reply:
I follow the riscv lower rules.
https://github.com/bytecodealliance/wasmtime/blob/f4b49b8c8942523b5bc44d09aa455fadaf848e41/cranelift/codegen/src/isa/riscv64/lower.isle#L1557-L1561

Reply:
I don't find a way to convert this `if-let` to `let`. Could you give me some suggestion? @elliottt

Reply:
It's a little more verbose, but:
```
(decl is_fneg_neg (IsFneg) u64)
(rule (is_fneg_neg (IsFneg.Result n _)) n)

(decl is_fneg_value (IsFneg) Value)
(rule (is_fneg_value (IsFneg.Result _ v)) v)

(decl pure is_fneg (Value) IsFneg)
(rule 1 (is_fneg (fneg x)) (IsFneg.Result 1 x))
(rule 0 (is_fneg x) (IsFneg.Result 0 x))

(rule (lower (has_type ty (fma x_src y_src z_src)))
  (let
    ((x_res IsFneg (is_fneg x_src))
     (y_res IsFneg (is_fneg y_src))
     (z_res IsFneg (is_fneg z_src))
     (neg_xy u64 (u64_xor (is_fneg_neg x_res) (is_fneg_neg y_res)))
     (x Value (is_fneg_value x_res))
     (y Value (is_fneg_value y_res))
     (z Value (is_fneg_value z_res)))
    (rv_fma ty neg_xy (is_fneg_neg z_res) x y z)))
```